File: gateway/engine.py
"""Where completions actually come from.

Two modes, chosen by environment:

  ONEROUTER_UPSTREAM_URL   an OpenAI-compatible base URL, e.g. https://api.openai.com/v1
  ONEROUTER_UPSTREAM_KEY   the bearer token for it
  ONEROUTER_UPSTREAM_MODEL optional; overrides the model id sent upstream

With both set, the gateway is a real proxy: your request is forwarded, the upstream's
tokens are streamed back, and you are billed on what came out. With them unset it falls
back to a local engine that answers deterministically. That fallback is NOT a language
model and never claims to be — it exists so the whole path (auth, routing, streaming,
billing) can be exercised with no credentials at all.
"""

from __future__ import annotations

import json
import os
import socket
import time
import urllib.error
import urllib.request

UPSTREAM_URL = os.environ.get("ONEROUTER_UPSTREAM_URL", "").rstrip("/")
UPSTREAM_KEY = os.environ.get("ONEROUTER_UPSTREAM_KEY", "")
UPSTREAM_MODEL = os.environ.get("ONEROUTER_UPSTREAM_MODEL", "")

LIVE = bool(UPSTREAM_URL and UPSTREAM_KEY)


class UpstreamError(Exception):
    """The upstream refused or could not be reached. Carries a status to pass on."""

    def __init__(self, status: int, message: str, *, timeout: bool = False):
        super().__init__(message)
        self.status = status
        self.message = message
        self.timeout = timeout


def count_tokens(text: str) -> int:
    """Deliberately approximate — four characters to a token. Real billing reads the
    upstream's usage block; this is the estimate used when it does not send one."""
    return max(1, len(text) // 4)


def _local_reply(model: dict, messages: list[dict]) -> str:
    last = next((m["content"] for m in reversed(messages) if m["role"] == "user"), "")
    turns = sum(1 for m in messages if m["role"] == "user")
    return (
        f"[local engine] No upstream is configured on this gateway, so this reply is "
        f"generated locally and is not model output.\n\n"
        f"Routed as:   {model['id']}\n"
        f"Host:        {model['hosts'][0]}\n"
        f"Context:     {model['context_length']:,} tokens\n"
        f"Turn:        {turns}\n"
        f"You said:    {last.strip()[:400] or '(nothing)'}\n\n"
        f"Everything around this text is real: the key was authenticated, the model was "
        f"resolved from the catalog, the response is streaming over SSE, and your balance "
        f"was charged at this model's catalog price. Set ONEROUTER_UPSTREAM_URL and "
        f"ONEROUTER_UPSTREAM_KEY to have a provider answer instead."
    )


def _chunks(text: str, size: int = 18):
    for i in range(0, len(text), size):
        yield text[i:i + size]


def stream_local(model: dict, messages: list[dict], max_tokens: int):
    """Yields text pieces, paced so the SSE path is exercised the way a real one is."""
    reply = _local_reply(model, messages)
    budget = max_tokens * 4 if max_tokens else len(reply)
    for piece in _chunks(reply[:budget]):
        yield piece
        time.sleep(0.012)


def stream_upstream(model: dict, messages: list[dict], max_tokens: int,
                    usage: dict | None = None, pin: list[str] | None = None):
    """Forwards to the configured provider and re-emits its content deltas.

    `usage` is filled in from the upstream's own counts when it reports them, which is
    what the request is then billed on — an estimate is only used when the upstream
    sends none.
    """
    payload = {
        "model": UPSTREAM_MODEL or model["id"],
        "messages": messages,
        "stream": True,
        # Ask for the usage block on the final chunk so billing is not guesswork.
        "stream_options": {"include_usage": True},
    }
    if max_tokens:
        payload["max_tokens"] = max_tokens
    if pin:
        # Pass a route pin through rather than pretending to honour it locally.
        payload["provider"] = {"only": pin}
    req = urllib.request.Request(
        f"{UPSTREAM_URL}/chat/completions",
        data=json.dumps(payload).encode(),
        headers={"Authorization": f"Bearer {UPSTREAM_KEY}",
                 "Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as res:
            for raw in res:
                line = raw.decode("utf-8", "replace").strip()
                if not line.startswith("data:"):
                    continue
                data = line[5:].strip()
                if data == "[DONE]":
                    return
                try:
                    frame = json.loads(data)
                except json.JSONDecodeError:
                    continue
                if usage is not None and frame.get("usage"):
                    counts = frame["usage"]
                    usage["prompt_tokens"] = counts.get("prompt_tokens") or 0
                    usage["completion_tokens"] = counts.get("completion_tokens") or 0
                if usage is not None and frame.get("provider"):
                    usage["provider"] = frame["provider"]
                try:
                    delta = frame["choices"][0]["delta"].get("content")
                except (KeyError, IndexError, TypeError):
                    continue
                if delta:
                    yield delta
    except urllib.error.HTTPError as err:
        raw = err.read().decode("utf-8", "replace")
        try:
            body = json.loads(raw)
            detail = (body.get("error") or {}).get("message") or raw[:200]
        except (json.JSONDecodeError, AttributeError):
            detail = raw[:200]
        raise UpstreamError(err.code, f"upstream {err.code}: {detail}") from err
    except (TimeoutError, socket.timeout) as err:
        raise UpstreamError(504, "no first token from the upstream within the window",
                            timeout=True) from err
    except urllib.error.URLError as err:
        # urllib wraps a socket timeout in URLError, so the reason has to be inspected.
        if isinstance(err.reason, (TimeoutError, socket.timeout)):
            raise UpstreamError(504, "no first token from the upstream within the window",
                                timeout=True) from err
        raise UpstreamError(502, f"upstream unreachable: {err}") from err
    except OSError as err:
        raise UpstreamError(502, f"upstream unreachable: {err}") from err


def stream(model: dict, messages: list[dict], max_tokens: int = 0,
           usage: dict | None = None, pin: list[str] | None = None):
    if LIVE:
        return stream_upstream(model, messages, max_tokens, usage, pin)
    return stream_local(model, messages, max_tokens)


def describe() -> str:
    if LIVE:
        return f"upstream {UPSTREAM_URL} (model {UPSTREAM_MODEL or 'passthrough'})"
    return "local engine (no upstream configured)"

File: web/lib/gateway/engine.ts
// Where completions actually come from. Ported from gateway/engine.py. With
// ONEROUTER_UPSTREAM_URL/_KEY set, the gateway proxies a real provider; unset, it falls
// back to a local engine that answers deterministically and says so in every reply.
import type { Model } from "../content/data";

export const UPSTREAM_URL = (process.env.ONEROUTER_UPSTREAM_URL || "").replace(/\/$/, "");
export const UPSTREAM_KEY = process.env.ONEROUTER_UPSTREAM_KEY || "";
export const UPSTREAM_MODEL = process.env.ONEROUTER_UPSTREAM_MODEL || "";
export const LIVE = Boolean(UPSTREAM_URL && UPSTREAM_KEY);

export interface Message {
  role: string;
  content: string;
}

export interface Usage {
  prompt_tokens?: number;
  completion_tokens?: number;
  provider?: string;
}

export class UpstreamError extends Error {
  status: number;
  timeout: boolean;
  constructor(status: number, message: string, opts: { timeout?: boolean } = {}) {
    super(message);
    this.status = status;
    this.timeout = opts.timeout ?? false;
  }
}

/** Deliberately approximate — four characters to a token. Real billing reads the
 * upstream's usage block; this is the estimate used when it sends none. */
export function countTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

function localReply(model: Model, messages: Message[]): string {
  const lastUser = [...messages].reverse().find((m) => m.role === "user");
  const last = lastUser?.content ?? "";
  const turns = messages.filter((m) => m.role === "user").length;
  const said = last.trim().slice(0, 400) || "(nothing)";
  return (
    `[local engine] No upstream is configured on this gateway, so this reply is ` +
    `generated locally and is not model output.\n\n` +
    `Routed as:   ${model.id}\n` +
    `Host:        ${model.hosts[0]}\n` +
    `Context:     ${model.context_length.toLocaleString("en-US")} tokens\n` +
    `Turn:        ${turns}\n` +
    `You said:    ${said}\n\n` +
    `Everything around this text is real: the key was authenticated, the model was ` +
    `resolved from the catalog, the response is streaming over SSE, and your balance ` +
    `was charged at this model's catalog price. Set ONEROUTER_UPSTREAM_URL and ` +
    `ONEROUTER_UPSTREAM_KEY to have a provider answer instead.`
  );
}

function* chunks(text: string, size = 18): Generator<string> {
  for (let i = 0; i < text.length; i += size) yield text.slice(i, i + size);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Yields text pieces, paced so the SSE path is exercised the way a real one is. */
export async function* streamLocal(model: Model, messages: Message[], maxTokens: number): AsyncGenerator<string> {
  const reply = localReply(model, messages);
  const budget = maxTokens ? maxTokens * 4 : reply.length;
  for (const piece of chunks(reply.slice(0, budget))) {
    yield piece;
    await sleep(12);
  }
}

interface ChatFrame {
  usage?: { prompt_tokens?: number; completion_tokens?: number };
  provider?: string;
  choices?: Array<{ delta?: { content?: string } }>;
}

/** Forwards to the configured provider and re-emits its content deltas. `usage` is
 * filled in from the upstream's own counts when it reports them — billing is not
 * guesswork unless the upstream sends none. */
export async function* streamUpstream(
  model: Model,
  messages: Message[],
  maxTokens: number,
  usage: Usage | null,
  pin: string[] | null
): AsyncGenerator<string> {
  const payload: Record<string, unknown> = {
    model: UPSTREAM_MODEL || model.id,
    messages,
    stream: true,
    // Ask for the usage block on the final chunk so billing is not guesswork.
    stream_options: { include_usage: true },
  };
  if (maxTokens) payload.max_tokens = maxTokens;
  if (pin) payload.provider = { only: pin }; // pass a route pin through, don't fake honoring it locally

  let res: Response;
  try {
    res = await fetch(`${UPSTREAM_URL}/chat/completions`, {
      method: "POST",
      headers: { Authorization: `Bearer ${UPSTREAM_KEY}`, "Content-Type": "application/json" },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(120_000),
    });
  } catch (err) {
    if ((err as Error).name === "TimeoutError") {
      throw new UpstreamError(504, "no first token from the upstream within the window", { timeout: true });
    }
    throw new UpstreamError(502, `upstream unreachable: ${(err as Error).message}`);
  }

  if (!res.ok) {
    const raw = await res.text();
    let detail = raw.slice(0, 200);
    try {
      const body = JSON.parse(raw);
      detail = body?.error?.message || detail;
    } catch {
      // raw text fallback already set
    }
    throw new UpstreamError(res.status, `upstream ${res.status}: ${detail}`);
  }
  if (!res.body) throw new UpstreamError(502, "upstream returned no body");

  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  let buf = "";
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buf.indexOf("\n")) !== -1) {
        const line = buf.slice(0, nl).trim();
        buf = buf.slice(nl + 1);
        if (!line.startsWith("data:")) continue;
        const data = line.slice(5).trim();
        if (data === "[DONE]") return;
        let frame: ChatFrame;
        try {
          frame = JSON.parse(data);
        } catch {
          continue;
        }
        if (usage && frame.usage) {
          usage.prompt_tokens = frame.usage.prompt_tokens || 0;
          usage.completion_tokens = frame.usage.completion_tokens || 0;
        }
        if (usage && frame.provider) usage.provider = frame.provider;
        const delta = frame.choices?.[0]?.delta?.content;
        if (delta) yield delta;
      }
    }
  } catch (err) {
    if ((err as Error).name === "TimeoutError") {
      throw new UpstreamError(504, "no first token from the upstream within the window", { timeout: true });
    }
    throw new UpstreamError(502, `upstream unreachable: ${(err as Error).message}`);
  }
}

export function stream(
  model: Model,
  messages: Message[],
  maxTokens = 0,
  usage: Usage | null = null,
  pin: string[] | null = null
): AsyncGenerator<string> {
  return LIVE ? streamUpstream(model, messages, maxTokens, usage, pin) : streamLocal(model, messages, maxTokens);
}

export function describe(): string {
  if (LIVE) return `upstream ${UPSTREAM_URL} (model ${UPSTREAM_MODEL || "passthrough"})`;
  return "local engine (no upstream configured)";
}
